src/algorithms/pid_lut/PIDLookup.cc

```diff
@@ -0,0 +1,142 @@
+// Copyright (C) 2024, Nathan Brei, Dmitry Kalinkin
+
+#include <algorithms/service.h>
+#include <edm4eic/MCRecoParticleAssociationCollection.h>
+#include <edm4eic/ReconstructedParticleCollection.h>
+#include <edm4hep/MCParticleCollection.h>
+#include <edm4hep/utils/vector_utils.h>
+#include <fmt/core.h>
+#include <podio/ObjectID.h>
+#include <cmath>
+#include <gsl/pointers>
+#include <stdexcept>
+
+#include "PIDLookup.h"
+#include "algorithms/pid_lut/PIDLookupConfig.h"
+#include "services/pid_lut/PIDLookupTableSvc.h"
+
+namespace eicrecon {
+
+void PIDLookup::init() {
+  auto& serviceSvc = algorithms::ServiceSvc::instance();
+  auto lut_svc = serviceSvc.service<PIDLookupTableSvc>("PIDLookupTableSvc");
+
+  m_lut = lut_svc->load(m_cfg.filename, {
+    .pdg_values=m_cfg.pdg_values,
+    .charge_values=m_cfg.charge_values,
+    .momentum_edges=m_cfg.momentum_edges,
+    .polar_edges=m_cfg.polar_edges,
+    .azimuthal_binning=m_cfg.azimuthal_binning,
+    .azimuthal_bin_centers_in_lut=m_cfg.azimuthal_bin_centers_in_lut,
+    .momentum_bin_centers_in_lut=m_cfg.momentum_bin_centers_in_lut,
+    .polar_bin_centers_in_lut=m_cfg.polar_bin_centers_in_lut,
+    .skip_legacy_header=m_cfg.skip_legacy_header,
+    .use_radians=m_cfg.use_radians,
+    .missing_electron_prob=m_cfg.missing_electron_prob,
+  });
+  if (m_lut == nullptr) {
+    throw std::runtime_error("LUT not available");
+  }
+}
+
+void PIDLookup::process(const Input& input, const Output& output) const {
+  const auto [recoparts_in, partassocs_in]          = input;
+  auto [recoparts_out, partassocs_out, partids_out] = output;
+
+  for (const auto& recopart_without_pid : *recoparts_in) {
+    edm4hep::MCParticle mcpart;
+    auto recopart = recopart_without_pid.clone();
+
+    // Find MCParticle from associations and propagate the relevant ones further
+    bool assoc_found = false;
+    for (auto assoc_in : *partassocs_in) {
+      if (assoc_in.getRec() == recopart_without_pid) {
+        if (assoc_found) {
+          warning("Found a duplicate association for ReconstructedParticle at index {}", recopart_without_pid.getObjectID().index);
+          warning("The previous MCParticle was at {} and the duplicate is at {}", mcpart.getObjectID().index, assoc_in.getSim().getObjectID().index);
+        }
+        assoc_found    = true;
+        mcpart         = assoc_in.getSim();
+        auto assoc_out = assoc_in.clone();
+        assoc_out.setRec(recopart);
+        partassocs_out->push_back(assoc_out);
+      }
+    }
+    if (not assoc_found) {
+      recoparts_out->push_back(recopart);
+      continue;
+    }
+
+    int true_pdg    = mcpart.getPDG();
+    int true_charge = mcpart.getCharge();
+    int charge      = recopart.getCharge();
+    double momentum = edm4hep::utils::magnitude(recopart.getMomentum());
+
+    double theta = edm4hep::utils::anglePolar(recopart.getMomentum()) / M_PI * 180.;
+    double phi   = edm4hep::utils::angleAzimuthal(recopart.getMomentum()) / M_PI * 180.;
```

Comment:
In the next iteration we should push for storing angles in radians in the LUT itself.

Reply:
We can. I think the degree interface is based on the original drich luts.